pkg/s3-proxy/utils/utils.go

```diff
@@ -136,3 +159,280 @@ func parseForwarded(forwarded string) (proto, host string) {
+// getDocumentFroARN retrieves a textual document from an AWS ARN for an S3 object, Secrets Manager secret, or
+// Systems Manager parameter.
+//
+// Note that S3 objects are usually supplied in S3 URL form (s3://bucket/key) instead, which is handled by
+// GetDocumentByURL directly.
+func getDocumentFromARN(rawURL string) ([]byte, error) {
+	docARN, err := arn.Parse(rawURL)
+	if err != nil {
+		return nil, err
+	}
+
+	switch docARN.Service {
+	case serviceS3:
+		if docARN.Region != "" {
+			return nil, fmt.Errorf("invalid S3 ARN: region cannot be set: %s", rawURL)
+		}
+
+		if docARN.AccountID != "" {
+			return nil, fmt.Errorf("invalid S3 ARN: account ID cannot be set: %s", rawURL)
+		}
+
+		parts := strings.SplitN(docARN.Resource, "/", 2) //nolint:gomnd // Splitting once
+		if len(parts) != 2 {                             //nolint:gomnd // Splitting once
+			return nil, fmt.Errorf("invalid S3 resource in ARN: %s", rawURL)
+		}
+
+		return getDocumentFromS3(parts[0], parts[1])
+
+	case serviceSecretsManager:
+		if docARN.Region == "" {
+			return nil, fmt.Errorf("invalid Secrets Manager ARN: region must be set: %s", rawURL)
+		}
+
+		if docARN.AccountID == "" {
+			return nil, fmt.Errorf("invalid Secrets Manager ARN: account ID must be set: %s", rawURL)
+		}
+
+		if !strings.HasPrefix(docARN.Resource, "secret:") {
+			return nil, fmt.Errorf("unsupported Secrets Manager resource in ARN: %s", rawURL)
+		}
+
+		return getDocumentFromSecretsManager(&docARN)
+
+	case serviceSSM:
+		if docARN.Region == "" {
+			return nil, fmt.Errorf("invalid SSM ARN: region must be set: %s", rawURL)
+		}
+
+		if docARN.AccountID == "" {
+			return nil, fmt.Errorf("invalid SSM ARN: account ID must be set: %s", rawURL)
+		}
+
+		if !strings.HasPrefix(docARN.Resource, "parameter/") {
+			return nil, fmt.Errorf("unsupported SSM resource in ARN: %s", rawURL)
+		}
+
+		return getDocumentFromSSM(&docARN)
+	}
+
+	return nil, fmt.Errorf("unsupported AWS service %#v in ARN: %s", docARN.Service, rawURL)
+}
+
+// getDocumentFromFile retrieves a textual document from a file.
+func getDocumentFromFile(filename string) ([]byte, error) {
+	return os.ReadFile(filename)
+}
+
+// getDocumentFromHTTP retrieves a textual document from an HTTP or HTTPS URL.
+func getDocumentFromHTTP(rawURL string) ([]byte, error) {
+	response, err := http.Get(rawURL) //nolint:gosec,noctx // We're getting HTTP data from an admin-specified URL.
+	if err != nil {
+		return nil, err
+	}
+
+	defer response.Body.Close()
+
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("request to %s failed with status code %d", rawURL, response.StatusCode)
+	}
+
+	return io.ReadAll(response.Body)
+}
+
+// getDocumentFromS3 retrieves a textual document from the specified S3 bucket and key.
+//
+// If the object is server-side encrypted, S3 will automatically decrypt this for us before returning it. Client-side
+// decryption is not supported.
+func getDocumentFromS3(bucket, key string) ([]byte, error) {
```

Comment:
S3-proxy is compatible with S3 endpoints. Can you make this part compatible too ?

Reply:
Ooh, that would be a good feature. Will put that in there too.

Reply:
I just remember another fact: s3-proxy support custom s3 endpoints and also the disable SSL option (see [here](https://github.com/oxyno-zeta/s3-proxy/blob/ea283f716a57ce18019d00cae939aa39f5079c4d/pkg/s3-proxy/s3client/manager.go#L74) ). I wonder if this won't need another object to be done and not just an url. If you have a good idea for this, I will probably take it for the global projet ;)

Reply:
Needed an additional object.